platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -477,58 +467,122 @@ public String toString() {
+            this(type, collectionFactory, valueCollector, Lists.newArrayList(value), 1);
         }
 
-        public CollectingSupplier(Collector<T> value) {
-            this(value, false);
+        // A constructor for sharing.
+        private CollectingSupplier(Class<C> type, Supplier<ImmutableCollection.Builder<T>> collectionFactory, ValueCollector<T> valueCollector, List<Collector<T>> collectors, int size) {
+            this.type = type;
+            this.collectionFactory = collectionFactory;
+            this.valueCollector = valueCollector;
+            this.collectors = collectors;
+            this.size = size;
+        }
+
+        @Override
+        protected Value<? extends C> calculateOwnValue(ValueConsumer consumer) {
+            return calculateValue(consumer);
+        }
+
+        @Nullable
+        @Override
+        public Class<C> getType() {
+            return type;
         }
 
         @Override
         public boolean calculatePresence(ValueConsumer consumer) {
-            return value.calculatePresence(consumer);
+            // We're traversing the elements in reverse addition order.
+            // When determining the presence of the value, the last argument wins.
+            // See also #collectExecutionTimeValues().
+            for (Collector<T> collector : Lists.reverse(getCollectors())) {
+                if (!collector.calculatePresence(consumer)) {
+                    // We've found an argument of add/addAll that is missing.
+                    // It makes the property missing regardless of what has been added before.
+                    // Because of the reverse processing order, anything that was added after it was just add/addAll that do not change the presence.
+                    return false;
+                }
+                if (isAbsentIgnoring(collector)) {
+                    // We've found an argument of append/appendAll, and everything added before it was present.
+                    // append/appendAll recovers the value of a missing property, so the property is also definitely present.
+                    return true;
+                }
+            }
+            // Nothing caused the property to become missing. There is at least one element by design, so the property is present.
+            assert size > 0;
+            return true;
         }
 
         @Override
         public Value<C> calculateValue(ValueConsumer consumer) {
             // TODO - don't make a copy when the collector already produces an immutable collection
             ImmutableCollection.Builder<T> builder = collectionFactory.get();
-            Value<Void> result = value.collectEntries(consumer, valueCollector, builder);
-            if (result.isMissing()) {
-                return result.asType();
+            Value<Void> compositeResult = Value.present();
+            for (Collector<T> collector : getCollectors()) {
+                if (compositeResult.isMissing() && !isAbsentIgnoring(collector)) {
```

Comment:
🤔 It seems that keeping the `absent ignoring` behavior makes things harder to understand and less efficient at runtime for little to no benefit or am I missing something?

Reply:
FTR: I gave it a try in this PR: https://github.com/gradle/gradle/pull/31555